client/asset/eth/eth.go

```diff
@@ -236,9 +238,71 @@ type ethFetcher interface {
 	signData(data []byte) (sig, pubKey []byte, err error)
 	syncProgress() ethereum.SyncProgress
 	transactionConfirmations(context.Context, common.Hash) (uint32, error)
+	getTransaction(context.Context, common.Hash) (*types.Transaction, int64, error)
 	txOpts(ctx context.Context, val, maxGas uint64, maxFeeRate *big.Int) (*bind.TransactOpts, error)
 	currentFees(ctx context.Context) (baseFees, tipCap *big.Int, err error)
 	unlock(pw string) error
+	getConfirmedNonce(context.Context, int64) (uint64, error)
+	transactionReceipt(context.Context, common.Hash) (*types.Receipt, error)
+}
+
+// monitoredTx is used to keep track of redemption transactions that have not
+// yet been confirmed. If a transaction has to be replaced due to the fee
+// being too low or another transaction being mined with the same nonce,
+// the replacement transaction's ID is recorded in the replacementTx field.
+type monitoredTx struct {
+	mtx sync.Mutex
+
+	tx             *types.Transaction
+	blockSubmitted int64
```

Comment:
The `tx` and `blockSubmitted` fields appear to be immutable after construction (and binary unmarshalling).  AFAICT, only `replacementTx` needs to be guarded by a mutex.

Reply:
What I mean that is our struct field grouping convention would be the following to express what fields need which concurrency controls:

```go
type monitoredTx struct {
	tx             *types.Transaction
	blockSubmitted uint64
	replacedTx     *common.Hash

	mtx          sync.Mutex
	replacementTx *common.Hash
}
```

However, looking at the current code, I see `replacedTx` is set after construction although that's not intuitive.  Is the reason that `(*assetWallet).Redeem` (called from `resubmitRedemption`) does not have the original tx to give to `w.monitorTx` where it's constructed?  Other than that technical reason, it the `replacedTx` field is immutable after being set, right?

Reply:
No you're right, `recordReplacementTx` is only called once. I've updated the struct and added some comments.